// simple-body-validator@1.0.2/lib/cjs/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.setTranslationObject = exports.setTranslationPath = exports.setFallbackLang = exports.setDefaultLang = exports.make = exports.ErrorBag = exports.Validator = exports.Password = exports.ImplicitRule = exports.Rule = void 0;
const validator_1 = __importDefault(require("./validator"));
exports.Validator = validator_1.default;
const lang_1 = __importDefault(require("./lang"));
const errorBag_1 = __importDefault(require("./validators/errorBag"));
exports.ErrorBag = errorBag_1.default;
const ruleContract_1 = __importDefault(require("./rules/ruleContract"));
const implicitRuleContract_1 = __importDefault(require("./rules/implicitRuleContract"));
const password_1 = __importDefault(require("./rules/password"));
class Rule extends ruleContract_1.default {
}
exports.Rule = Rule;
;
class ImplicitRule extends implicitRuleContract_1.default {
}
exports.ImplicitRule = ImplicitRule;
;
class Password extends password_1.default {
}
exports.Password = Password;
;
function make(data = {}, rules = {}, customMessages = {}, customAttributes = {}) {
    return new validator_1.default(data, rules, customMessages, customAttributes);
}
exports.make = make;
;
function setDefaultLang(lang) {
    lang_1.default.setDefaultLang(lang);
}
exports.setDefaultLang = setDefaultLang;
;
function setFallbackLang(lang) {
    lang_1.default.setFallbackLang(lang);
}
exports.setFallbackLang = setFallbackLang;
;
function setTranslationPath(path) {
    lang_1.default.setPath(path);
}
exports.setTranslationPath = setTranslationPath;
;
function setTranslationObject(translations) {
    lang_1.default.setTranslationObject(translations);
}
exports.setTranslationObject = setTranslationObject;
;
__exportStar(require("./rules/registerRule"), exports);
__exportStar(require("./rule"), exports);
