// simple-body-validator/lib/ecm/index.js
import Validator from './validator';
import Lang from './lang';
import ErrorBag from './validators/errorBag';
import RuleContract from './rules/ruleContract';
import ImplicitRuleContract from './rules/implicitRuleContract';
import { default as PasswordRule } from './rules/password';
class Rule extends RuleContract {
}
;
class ImplicitRule extends ImplicitRuleContract {
}
;
class Password extends PasswordRule {
}
;
function make(data = {}, rules = {}, customMessages = {}, customAttributes = {}) {
    return new Validator(data, rules, customMessages, customAttributes);
}
;
function setDefaultLang(lang) {
    Lang.setDefaultLang(lang);
}
;
function setFallbackLang(lang) {
    Lang.setFallbackLang(lang);
}
;
function setTranslationPath(path) {
    Lang.setPath(path);
}
;
function setTranslationObject(translations) {
    Lang.setTranslationObject(translations);
}
;
export { Rule, ImplicitRule, Password, Validator, ErrorBag, make, setDefaultLang, setFallbackLang, setTranslationPath, setTranslationObject, };
export * from './rules/registerRule';
export * from './rule';
